src/storage/store/mempool_poller.rs

```diff
@@ -0,0 +1,149 @@
+}
+
+impl MempoolMessage {
+    pub fn fid(&self) -> u64 {
+        match self {
+            MempoolMessage::UserMessage(msg) => msg.fid(),
+            MempoolMessage::ValidatorMessage(msg) => msg.fid(),
+        }
+    }
+
+    pub fn to_proto(&self) -> proto::MempoolMessage {
+        let msg = match self {
+            MempoolMessage::UserMessage(msg) => {
+                proto::mempool_message::MempoolMessage::UserMessage(msg.clone())
+            }
+            _ => todo!(),
+        };
+        proto::MempoolMessage {
+            mempool_message: Some(msg),
+        }
+    }
+}
+
+impl MempoolPoller {
+    pub fn for_read_node(&self) -> bool {
+        self.messages_request_tx.is_none()
+    }
+    fn time_with_shard(&self, key: &str, value: u64) {
+        let key = format!("mempool_poller.{}", key);
+        self.statsd_client
+            .time_with_shard(self.shard_id, key.as_str(), value);
+    }
+
+    pub(crate) async fn pull_messages(
+        &mut self,
+        max_wait: Duration,
+    ) -> Result<Vec<MempoolMessage>, MempoolPollerError> {
+        if let Some(messages_request_tx) = &self.messages_request_tx {
+            let now = std::time::Instant::now();
+            let (message_tx, message_rx) = oneshot::channel();
+
+            if let Err(err) = messages_request_tx
+                .send(MempoolMessagesRequest {
+                    shard_id: self.shard_id,
+                    message_tx,
+                    max_messages_per_block: self.max_messages_per_block,
+                })
+                .await
+            {
+                error!(
+                    "Could not send request for messages to mempool {}",
+                    err.to_string()
+                )
+            }
+
+            match timeout(max_wait, message_rx).await {
+                Ok(response) => match response {
+                    Ok(new_messages) => {
+                        let elapsed = now.elapsed();
+                        self.time_with_shard("pull_messages", elapsed.as_millis() as u64);
+                        Ok(new_messages)
+                    }
+                    Err(err) => Err(err)?,
+                },
+                Err(_) => {
+                    error!("Did not receive messages from mempool in time");
+                    // Just proceed with no messages
+                    Ok(vec![])
+                }
+            }
+        } else {
+            Ok(vec![])
+        }
+    }
+
+    // Groups messages by fid and creates a transaction for each fid
+    pub fn create_transactions_from_mempool(
+        &mut self,
+        messages: Vec<MempoolMessage>,
+    ) -> Result<Vec<Transaction>, MempoolPollerError> {
+        let mut transactions = vec![];
+
+        let grouped_messages = messages.iter().into_group_map_by(|msg| msg.fid());
+        let unique_fids = grouped_messages.keys().len();
+        for (fid, messages) in grouped_messages {
+            let mut transaction = Transaction {
```

Comment:
This living outside the engine and having the logic split between here and the engine feels weird. Again, might be fine, just wanted to make a node. I think the block engine might have different rules on which messages to include (e.g. no user messages except storage lend) so this split might be necessary, still feels a little weird.

Reply:
I'm not really sure why we're handling storage being missing in this way. I think we should be running this check inside  replay_snapchain_transaction the same way we check for active signer there.